dspace-api/src/main/java/org/dspace/google/GoogleAsyncEventListener.java

```diff
@@ -0,0 +1,236 @@
+import org.dspace.core.Constants;
+import org.dspace.core.Context;
+import org.dspace.services.ConfigurationService;
+import org.dspace.services.model.Event;
+import org.dspace.usage.AbstractUsageEventListener;
+import org.dspace.usage.UsageEvent;
+import org.springframework.beans.factory.annotation.Autowired;
+
+
+/**
+ * Notifies Google Analytics of Bitstream VIEW events. These events are stored in memory and then
+ * asynchronously processed by a single seperate thread.
+ *
+ * @author April Herron
+ */
+public class GoogleAsyncEventListener extends AbstractUsageEventListener {
+
+    private static final int MAX_TIME_SINCE_EVENT = 14400000;
+    private static final String ANALYTICS_BATCH_ENDPOINT = "https://www.google-analytics.com/batch";
+    private static Logger log = Logger.getLogger(GoogleAsyncEventListener.class);
+    private static String analyticsKey;
+    private static CloseableHttpClient httpclient;
+    private static Buffer buffer;
+    private static ExecutorService executor;
+    private static Future future;
+    private static boolean destroyed = false;
+
+    @Autowired(required = true)
+    ConfigurationService configurationService;
+
+    @PostConstruct
+    public void init() {
+        analyticsKey = configurationService.getProperty("google.analytics.key");
+        if (StringUtils.isNotEmpty(analyticsKey)) {
+            int analyticsBufferlimit = configurationService.getIntProperty("google.analytics.buffer.limit", 256);
+            buffer = BufferUtils.synchronizedBuffer(new CircularFifoBuffer(analyticsBufferlimit));
+            httpclient = HttpClients.createDefault();
+            executor = Executors.newSingleThreadExecutor();
+            future = executor.submit(new GoogleAnalyticsTask());
+        }
+    }
+
+    @Override
+    public void receiveEvent(Event event) {
+        if ((event instanceof UsageEvent)) {
+            if (StringUtils.isNotEmpty(analyticsKey)) {
+                UsageEvent ue = (UsageEvent) event;
+                log.debug("Usage event received " + event.getName());
+                try {
+                    if (ue.getAction() == UsageEvent.Action.VIEW &&
+                            ue.getObject().getType() == Constants.BITSTREAM) {
+
+                        // Client Id, should uniquely identify the user or device. If we have a session id for the user
+                        // then lets use it, else generate a UUID.
+                        String cid;
+                        if (ue.getRequest().getSession(false) != null) {
+                            cid = ue.getRequest().getSession().getId();
+                        } else {
+                            cid = UUID.randomUUID().toString();
+                        }
+                        buffer.add(new GoogleAnalyticsEvent(cid, getIPAddress(ue.getRequest()), ue.getRequest()
+                                .getHeader("USER-AGENT"), ue.getRequest().getHeader("referer"), ue.getRequest()
```

Comment:
Instead of only using the `referer` header, we may want to first check if our custom `X-REFERRER` header exists...if it doesn't, then fallback to using `referer`.  That'd also align better with our logs, which could let you see the same requests in google analytics & in dspace.log.  See again https://github.com/DSpace/DSpace/pull/3303